### ISSUE TITLE
Reading write-only PPI port 0xAB results in 0x82 always (it seems) - not so on real MSX

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1505]
As Accumulator wrote:
> If I check PPI port &HAB (which should be WO), on a real MSX, gives 0FFH, on OpenMSX the value 082H..
> Is it only my real MSX or also other real MSX-es?
> Whatever value I write to this port, real MSX always 0FFh (in my case)...OpenMSX always 082H...
> I do not know if others, on real MSX, when checking port 0ABH, give 0FFH or other value.
(See https://www.msx.org/comment/reply/65515/442439)

Would indeed be interested to see what this gives on different kinds of MSX machines: `?HEX$(INP(&HAB))`
I get `0xF1` on my Philips NMS 8250 (S3527) and `0xFF` on my Panasonic FS-A1GT (T9769C).

As Wouter wrote to me:

> A8, A9, AA can be programmed to be either IN or OUT, so in the xml they must be IO. It's true that in MSX they only have one fixed direction (set early during the boot sequence, and then never changed again).
> i'll still check port AB ...
> ... the PPI (82C55A) datasheet mentions that the control port (mapped on IO port 0xAB in MSX) CAN be read back. For MSX that isn't strictly needed, so maybe the integration in (some?) MSX-engines doesn't support this read-back functionality

So, it would be especially interesting to see what happens on MSX machines without MSX engine IC.

Did some more tests:
Mitsubishi ML-FX1 with S3527 also gave 0xF1.
Canon V-20 (without any engine) gave 0xFA.

In any case, it never returned `0x82` like in openMSX.

So, what the heck is happening on real MSXes regarding this port?

**Comment by EMLommers** [comment]
Maybe it is related to the engine.  
No Engine - 0xFA
S3527 - 0xF1
S1985 - 0xFF
T9769A/B/C - 0xFF

Also, when doing init the value &H82 is written to &HAB
After init doing OUT &HAB,&H82 (real MSX SONY HBF) it hangs, OpenMSX does not hang!
Am I the only one?

**Comment by 0cjs** [comment]
Yes, it is related to the engine, but possibly a bunch of other things too, incuding the PCB design and what peripherals are on the MSX data bus, including any cartridges that may be plugged in. In the thread mentioned above I did a bit of analysis [here](https://msx.org/forum/msx-talk/hardware/io-port-mirroring-please-help?page=6#comment-443390); and in earlier messages. This is essentialy a copy of what I wrote there:

----------

I just had a quick [at the 8255 datasheet](https://web.archive.org/web/20130104230953/http://bitsavers.informatik.uni-stuttgart.de/pdf/intel/MCS80/98-153B_Intel_8080_Microcomputer_Systems_Users_Manual_197509.pdf) (the 8255 data starts on page 5-113, PDF page 181), and realised that you're reading from port $AB which is likely not connected. The datasheet, on page 5-115 (PDF page 183) clearly says, "The Control Word Register can Only be written into. No Read operation of the Control Word Register is allowed." So if the designers of the particular system you're using went the safe route on their address decoding, they do not enable the PPI for reads from that address and it's not connected. If they have even connected it, the behaviour would be undefined, so there's nothing to emulate, really. (What happens when you read in that case may vary even between different 8255 chips, not to mention between systems that use an actual 8255 vs. ones that have the functionality integrated into another chip.)

And BTW, when I see such a strong "not allowed" statement in a datasheet, I would seriously consider ensuring that the chip is not enabled for that situation. When doing such not allowed stuff it's not unusual for chips to get into weird non-working states that can't be fixed without a power cycle, and conceivably there's even the possibility of permanent damage. (Though I would hope the chip designers put in their own safety mechanisms to prevent that sort of thing.)

----------

According to the ASCII MSX 2 Technical Handbook [page 390](https://archive.org/details/MSX2TechnicalHandBookFE1986/page/n393/mode/1up), $A8-$AB are for a "Parallel Port (8255)", which implies that the defined behaviour is the same as an 8255. And port $AB is the "mode port" which for an 8255, as I mentioned in the data sheet above, must not be read, which means that reading it results in undefined behaviour. (The old Sony MSX Technical Handbook claims on [page 41](https://archive.org/details/MSXTechnicalHandbookBySony/page/n43/mode/1up?view=theater) that $AB is R/W, but this is surely a typo.)

So this basically means that manufacturers are allowed to do whatever they heck they want on read of that port, which has the following consequences for emulation:

1. There is no generic emulation for it; what your emulation does, if it's going to be "correct," has to be individually determined for each MSX model.

2. For MSX models where it's determined that the implementation _**does**_ write the data bus on a read request for that port, you need to determine if the chip has defined behaviour for that. If so, you're done, otherwise you need to do a statistical sampling of those chips across a wide range of dates to determine what its undocumented behaviour is and whether it's consistent between all chips.

3. For MSX models where it's determined that the implementation _**does not**_ write the data bus on a read request for that port, the behaviour will be determined by the PCB (possibly changing for different manufacturing runs of that PCB) and everything connected to the data bus, since that will change the bus capacitance. Further, what value you will read will vary over the amount of time it's been since the last device (or pull-ups/downs) drove the data bus. Getting this right probably involves a huge amount of statistical sampling, probably starting with pulling the Z80 and driving the bus yourself with a custom device that can generate the read requests and read the values you get over time after that without anything (such as reading another instruction for the Z80) driving the bus. And of course this will vary from configuration to configuration even on the same machine, depending on e.g. what's plugged into the cartridge port(s).

----------

Regarding the Sony "HBF" (It's not clear exactly which model you mean) hanging when writing $82 to the control port ($AB), I've verified that the configuration appears correct:

```
Control Word ($82 = 01000010)

    bit val description
    ─────────────────────────────────────────────────────────────────
     7   0  Mode set flag: 0=inactive (Mode 0: basic input/output)
    6:5 10  Group A mode selection: 00=Mode 0 (as above)
     4   0  Port A: 0=output
     3   0  Port C upper: 0=output
     2   0  Group B mode selection: 0=Mode 0 (as above)
     1   1  Port B: 1=input
     0   0  Port C lower: 0=output
```

However, the datasheet (page 5-116, PDF page 184) says:

> All of the output registers, including the status flip-flops, will be reset whenever the mode is changed.

Though what you're writing isn't technically a mode _change_ (since you're asking for the same mode it's already in), it's possible that this actually means (or was taken by some clone-8255 designers to mean) that any write to the control register will reset all of the output registers. This would reset the page mappings to primary slot 0 for all four pages which, on a machine such as the [Sony HB-F1](https://www.msx.org/wiki/Sony_HB-F1#Slot_Map) would unmap the RAM (which is in slot 3) from pages 2 and 3, leaving instead unmapped address space from slot 0 there. This would cause the entire housekeeping data area at the top of page 3 to become unmapped, which almost certainly immediately crash the system.

**Comment by EMLommers** [comment]
Manuel wants perfect emjulation, therefore also the results of those ports.....
It as exactly what you earlier wrote,  and I can just agree....

@Ocjs it is of very import to them otherwise would Manuel not have  opened this topic.

I have also check the c8255, MSX engine and the z80 familiy data book from 89 (intersesting one btw)

**Comment by 0cjs** [comment]
Yeah, the issue there is, "which machine do you want to emulate?" Not even as in, "Do you want to emulate an HB-F1 or HB-F1II?" but "I have two Sony HB-F1 machines in front of me that behave differently in this respect; which one should it act like?"

**Comment by EMLommers** [comment]
You can for the fun ? hex$(inp(&HAB))
and see if they all provide the same results.  (They have S3257 Enigne, S1985 or None)
OpenMSX like to achieve perfect emulation, and that includes the ports

But what I like to code is the Machine configurator... and that is escpecially interseting,  
You can Select any MSX-Engine, with any base mainboard, RTC yes or no, etc etc

**Comment by EMLommers** [comment]
@0cjs, HBF is HB-F700D. I could have been more specific.
